org.eclipse.transformer/src/main/java/org/eclipse/transformer/util/FileUtils.java

```diff
@@ -36,89 +44,215 @@ private FileUtils() {}
+
+		ByteBuffer finalBuffer = read(inputName, inputStream, initialBuffer, requested);
+
+		if ( finalBuffer != initialBuffer ) {
+			inputBuffer.setInputBuffer(finalBuffer);
 		}
 
-		// System.out.println("Count [ " + count + " ] Adjusted to [ " +
-		// intCount + " ]");
+		if ( requested < 0 ) {
+			logger.debug("Read [ {} ] bytes [ {} ]", inputName, finalBuffer.limit());
+		}
 
-		return count;
+		return new ByteDataImpl(inputName, finalBuffer);
+	}
+
+	public static ByteData read(Logger logger, String inputName, InputStream inputStream, InputBuffer inputBuffer)
+		throws IOException {
+		return read(logger, inputName, inputStream, -1, inputBuffer);
 	}
 
 	/**
-	 * Read data from an input stream into a buffer. Allocate a new buffer if
-	 * the parameter buffer is too small for the requested read.
+	 * Read bytes from an input stream.
+	 * <p>
+	 * Place the read bytes in the byte buffer.  Allocate a new buffer if the initial buffer
+	 * is too small.
+	 * <p>
+	 * If the requested count of bytes is less than zero, read all available bytes from the input stream.  Otherwise, read the requested number of bytes.
+	 * <p>
+	 * Exceptions are thrown if a read fails, or if too few bytes were read, or if an array could not
+	 * be allocated large enough for all bytes which were to be read.
+	 * <p>
+	 * An exception is thrown if the requested number of bytes is too large.
 	 *
-	 * @param inputName A name associated with the input stram.
-	 * @param inputStream The stream from which to read bytes.
-	 * @param buffer A buffer into which to place the read bytes.
-	 * @param count The number of bytes to read.
-	 * @return The final buffer which contains the read bytes. The same as the
-	 *         parameter buffer if that buffer had a sufficient capacity to read
-	 *         the requested count of bytes. A new buffer if the parameter
-	 *         buffer was too small.
-	 * @throws IOException Thrown if an error occurred during a read.
+	 * @param inputName The name of the input stream.
+	 * @param inputStream A stream to be read.
+	 * @param buffer A byte buffer.
+	 * @param requested The count of bytes to read from the stream. If less than
+	 *            zero, all available bytes will be read.
+	 * @return Byte data from the read.
+	 * @throws IOException Thrown if a read fails, or if overflow or underflow
+	 *             occurs.
 	 */
-	public static ByteBuffer read(String inputName, InputStream inputStream, ByteBuffer buffer, int count)
+	public static ByteBuffer read(String inputName, InputStream inputStream, ByteBuffer buffer, int requested)
 		throws IOException {
-		if (count != -1) {
-			verifyArray(0, count);
-			if (count > buffer.capacity()) {
-				buffer = ByteBuffer.allocate(count);
-			}
+
+		if ( requested == 0 ) {
+			buffer.position(0);
+			buffer.limit(0);
+
+			return buffer;
 		}
-		buffer.clear();
-		for (int bytesRead; (bytesRead = inputStream.read(buffer.array(), buffer.arrayOffset() + buffer.position(),
```

Comment:
Why is this all rewritten? I assume it is because of the bad changes in ByteDataImpl which does not duplicate the buffer.